edgehog-device-runtime-containers/src/docker/container.rs

```diff
@@ -0,0 +1,888 @@
+    },
+    errors::Error as BollardError,
+    models::{
+        ContainerInspectResponse, EndpointSettings, HostConfig, PortBinding, RestartPolicy,
+        RestartPolicyNameEnum,
+    },
+};
+use tracing::{debug, info, instrument, trace, warn};
+
+use crate::{
+    client::*,
+    requests::{container::parse_port_binding, BindingError},
+};
+
+/// Error for the container operations.
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ContainerError {
+    /// couldn't create the container
+    Create(#[source] BollardError),
+    /// couldn't inspect the container
+    Inspect(#[source] BollardError),
+    /// couldn't remove the container
+    Remove(#[source] BollardError),
+    /// couldn't start container.
+    Start(#[source] BollardError),
+    /// couldn't stop container
+    Stop(#[source] BollardError),
+    /// missing image reference in container definition
+    Image,
+}
+
+/// Docker container struct.
+#[derive(Debug, Clone, Eq)]
+pub struct Container<S>
+where
+    S: Hash + Eq,
+{
+    /// Id of the docker container.
+    pub id: Option<String>,
+    /// Assign the specified name to the container.
+    ///
+    /// Must match /?[a-zA-Z0-9][a-zA-Z0-9_.-]+.
+    pub name: S,
+    /// The name (or reference) of the image to use.
+    ///
+    /// The id of the image is optional since it will be available only when the image is.
+    pub image: S,
+    /// Network to link the container with.
+    pub networks: Vec<S>,
+    /// The hostname to use for the container.
+    ///
+    /// Defaults to the container name.
+    pub hostname: Option<S>,
+    /// The behaviour to apply when the container exits.
+    ///
+    /// See the [create container
+    /// API](https://docs.docker.com/engine/api/v1.43/#tag/Container/operation/ContainerCreate) for
+    /// possible values.
+    pub restart_policy: RestartPolicyNameEnum,
+    /// A list of environment variables to set inside the container.
+    ///
+    /// In the form of `NAME=VALUE`.
+    pub env: Vec<S>,
+    /// A list of volume bindings for this container.
+    pub binds: Vec<S>,
+    /// Describes the mapping of container ports to host ports.
+    ///
+    /// It uses the container's port-number and protocol as key in the format `<port>/<protocol>`, for
+    /// example, 80/udp.
+    pub port_bindings: PortBindingMap<S>,
+    /// Gives the container full access to the host.
+    ///
+    /// Defaults to false.
+    pub privileged: bool,
+}
+
+impl<S> Container<S>
+where
+    S: Hash + Eq,
+{
+    /// Create a new container.
+    pub fn new(name: S, image: S) -> Self {
+        Self {
+            id: None,
+            name,
+            image,
+            hostname: None,
+            restart_policy: RestartPolicyNameEnum::EMPTY,
+            env: Vec::new(),
+            binds: Vec::new(),
+            networks: Vec::new(),
+            port_bindings: PortBindingMap::new(),
+            privileged: false,
```

Comment:
You could use ...Default::default() for all the fields apart from name, image and restart_policy

Reply:
No because the struct is not Default

Reply:
Oh you're right! I thought the struct update sintax could be used with Default::default also in case the missing fields implemented Default, not the whole struct